### PR TITLE
fix: disconnect logic for EIP6963 & Injected provider types for @web3modal/ethers

### LINKED ISSUE #2288: [bug] Modal doesn't properly disconnect when disconnecting - @web3modal/ethers
### Link to minimal reproducible example

https://lab.web3modal.com/library/ethers/

### Summary

The `@web3modal/ethers` library is not handling the disconnect logic correctly. When selecting MetaMask from the modal, the MetaMask extension does not pop up, and it auto-connects the wallet instead. 

### Steps to Reproduce
1. **Open the WalletConnect Modal**:
   - Launch the modal to connect to a wallet.

2. **Select MetaMask**:
   - Select MetaMask from the modal options.
3. **Select a Wallet**:
   - Select a Wallet from the Metamask Pop up .

4. **Disconnect and Try to Connect Again**:
   - Attempt to disconnect the MetaMask wallet.
   - Open the modal again to select MetaMask.
   - The wallet should connect automatically.

### Expected Behavior
Each time the WalletConnect modal is opened, it should prompt the MetaMask extension to pop up and allow the user to select the wallet they want to connect to. This should happen every time the modal is launched, ensuring the user can switch between different wallets as needed.

### Actual Behavior
When selecting MetaMask, the extension does not pop up, and the wallet auto-connects. This prevents the user from selecting a different wallet each time, which is necessary for certain use cases.

### List of related npm package versions

@web3modal/ethers

### PR DESCRIPTION
**Pull request by hmzakhalid**
# Changes
- fix: disconnect logic for EIP6963 & Injected provider types in packages/ethers

# Associated Issues

closes #2288

### CODE REVIEW

**Comment by hmzakhalid** [comment]
Hey @tomiir, Can you take a look at this Please?

**Comment by tomiir** [comment]
hey @hmzakhalid ! ty for the contribution. 
I'm bumping this up with the rest of the team as not sure we want to add more clicks to the connection flow. 
I'm thinking maybe a proper way to handle this would be with a config flag on the ethers configuration that would dictate if we should or should not revoke the permissions. Agree there are cases where this would be a better ux, but would like input from @glitch-txs @enesozturk  as well. 

Code looks good overall, would try-catch the requests to the wallet in case they are not supported to prevent unexpected behaviors but otherwise looks good!

File: packages/ethers/src/client.ts [hunk 1]
```diff
@@ -316,6 +313,39 @@ export class Web3Modal extends Web3ModalScaffold {
...
+          }[]
+          const ethAccountsPermission = permissions.find(
+            permission => permission.parentCapability === 'eth_accounts'
+          )
+
+          if (ethAccountsPermission) {
+            await provider.request({
+              method: 'wallet_revokePermissions',
+              params: [{ eth_accounts: {} }]
+            })
```
[the remark this answers is not in the document]
Comment (hmzakhalid) [reply packages/ethers/src/client.ts]: added the try-catch block. but also not sure if we need the, I assumed all wallets would support these methods.

**Comment by hmzakhalid** [comment]
> hey @hmzakhalid ! ty for the contribution. I'm bumping this up with the rest of the team as not sure we want to add more clicks to the connection flow. I'm thinking maybe a proper way to handle this would be with a config flag on the ethers configuration that would dictate if we should or should not revoke the permissions. Agree there are cases where this would be a better ux, but would like input from @glitch-txs @enesozturk as well.
> 
> Code looks good overall, would try-catch the requests to the wallet in case they are not supported to prevent unexpected behaviors but otherwise looks good!

Thank you for the feedback and for bringing this up with the team. I'd like to point out that this should be the default behavior when the disconnect button is clicked. 
- Users expect that clicking a disconnect button will fully disconnect their wallet from the site, including revoking any permissions. This is a standard behavior that users are familiar with, and it aligns with their expectations. 
- The `wagmi` connector, for instance, ensures that the wallet is fully disconnected from the site. 
- If the wallet is not fully disconnected, it can lead to confusion and frustration, as users might not understand why they are still connected or why the wallet does not prompt them to reconnect properly. (As it happened in my case).

Would love feedback from the rest of the team.

**Comment by tomiir** [comment]
Been re-reading this. I'm on board. Verified and it matches wagmi's behavior. 
Would love to get confirmation from team and we could move ahead with it 💪

**Comment by hmzakhalid** [comment]
Hey @svenvoskamp , Could you please take a look at this PR as well?

**Review by svenvoskamp — APPROVED** [review]
LGTM, could you please fix lint issues?

**Review by tomiir — APPROVED** [review]
🚀

File: packages/ethers/src/client.ts [hunk 2]
```diff
@@ -319,6 +316,47 @@ export class Web3Modal extends Web3ModalScaffold {
           // eslint-disable-next-line no-negated-condition
         } else if (providerType === ConstantsUtil.AUTH_CONNECTOR_ID) {
           await this.authProvider?.disconnect()
+        } else if (providerType === ConstantsUtil.EIP6963_CONNECTOR_ID && provider) {
+          try {
+            const permissions = (await provider.request({ method: 'wallet_getPermissions' })) as {
```
Comment (enesozturk) [inline packages/ethers/src/client.ts:321]: Don't we have the types of the response here? Why linter showing that assertion is unnecessary? If we don't know the return type, maybe we should override the lint for these lines? @tomiir @svenvoskamp

Comment (hmzakhalid) [reply packages/ethers/src/client.ts:321]: Nope, the return type is a generic Promise<any> so ts can't infer the exact type of the response.
```js
request(req: W3mFrameTypes.RPCRequest): Promise<any>;
```

Also, i dont see the linting issue. its TS complaining in the line below `'permissions' is of type 'unknown'.`

Comment (hmzakhalid) [reply packages/ethers/src/client.ts:321]: Nvm, I just ran the linter and saw the issue

Comment (hmzakhalid) [reply packages/ethers/src/client.ts]: Fixed it, will wait for others to reply before resolving the conversation.

**Comment by hmzakhalid** [comment]
Shouldn't this disconnect function also be updated? i suppose this is used in the `useDisconnect` hook
https://github.com/hmzakhalid/web3modal/blob/V4/packages/ethers/src/client.ts#L616-L637

@enesozturk @svenvoskamp @tomiir

**Comment by enesozturk** [comment]
> Shouldn't this disconnect function also be updated? i suppose this is used in the `useDisconnect` hook https://github.com/hmzakhalid/web3modal/blob/V4/packages/ethers/src/client.ts#L616-L637
> 
> @enesozturk @svenvoskamp @tomiir

@hmzakhalid That's a good point - yes. The public `disconnect()` function is for hooks and the JS client and should do the same thing as `connectionControllerClient.disconnect` does.

**Comment by hmzakhalid** [comment]
> > Shouldn't this disconnect function also be updated? i suppose this is used in the `useDisconnect` hook https://github.com/hmzakhalid/web3modal/blob/V4/packages/ethers/src/client.ts#L616-L637
> > @enesozturk @svenvoskamp @tomiir
> 
> @hmzakhalid That's a good point - yes. The public `disconnect()` function is for hooks and the JS client and should do the same thing as `connectionControllerClient.disconnect` does.

Thought so, pushed the updated code.

**Comment by tomiir** [comment]
Final request: given we are repeating the same thing 3 times, can we abstract it to a method/util?
Basically something like `disconnectProvider({ provider, providerType }) => void` 🙏 
Else looks good 🖖 🚀

**Review by glitch-txs — APPROVED** [review]

**Comment by hmzakhalid** [comment]
Bumping this Up

**Comment by hmzakhalid** [comment]
Bumping this up @glitch-txs @tomiir

**Pull request merged**